src/bridge.rs

```diff
@@ -118,131 +116,238 @@ impl CommandLight {
     }
 }
 
+/// An unauthenticated bridge is a bridge that has not
+#[derive(Debug, Clone)]
+pub struct UnauthBridge {
+    /// The IP-address of the bridge.
+    pub ip: std::net::IpAddr,
+    pub(self) client: reqwest::blocking::Client,
+}
+
+impl UnauthBridge {
+    /// Consumes the bidge and return a new one with a configured username.
+    /// ### Example
+    /// ```rust
+    /// let bridge = hueclient::Bridge::for_ip([192u8, 168, 0, 4])
+    ///    .with_user("rVV05G0i52vQMMLn6BK3dpr0F3uDiqtDjPLPK2uj");
+    /// ```
+    pub fn with_user(self, username: impl Into<String>) -> Bridge {
+        Bridge {
+            ip: self.ip,
+            username: username.into(),
+            client: self.client,
+        }
+    }
+
+    /// This function registers a new user at the provided brige, using `devicetype` as an
+    /// identifier for that user. It returns an error if the button of the bridge was not pressed
+    /// shortly before running this function.
+    /// ### Example
+    /// ```rust
+    /// let mut bridge = hueclient::Bridge::for_ip([192u8, 168, 0, 4]);
+    /// let password = bridge.register_user("mylaptop").unwrap();
+    /// // now this password can be stored and reused
+    /// ```
+    pub fn register_user(self, devicetype: &str) -> crate::Result<Bridge> {
+        #[derive(Serialize)]
+        struct PostApi {
+            devicetype: String,
+        }
+        #[derive(Debug, Deserialize)]
+        struct Username {
+            username: String,
+        }
+        let obtain = PostApi {
+            devicetype: devicetype.to_string(),
+        };
+        let url = format!("http://{}/api", self.ip);
+        let resp: BridgeResponse<SuccessResponse<Username>> =
+            self.client.post(&url).json(&obtain).send()?.json()?;
+        let resp = resp.get()?;
+
+        Ok(Bridge {
+            ip: self.ip,
+            username: resp.success.username,
+            client: self.client,
+        })
+    }
+}
+
+/// The bridge is the central access point of the lamps is a Hue setup, and also the central access
+/// point of this library.
 #[derive(Debug)]
 pub struct Bridge {
+    /// The IP-address of the bridge.
     pub ip: std::net::IpAddr,
-    pub username: Option<String>,
-    client: reqwest::blocking::Client,
+    /// This is the username of the currently logged in user.
+    pub username: String,
+    pub(self) client: reqwest::blocking::Client,
 }
 
 impl Bridge {
-    pub fn for_ip(ip: std::net::IpAddr) -> Bridge {
-        Bridge {
-            ip,
-            username: None,
+    /// Create a bridge at this IP. If you know the IP-address, this is the fastest option. Note
+    /// that this function does not validate whether a bridge is really present at the IP-address.
+    /// ### Example
+    /// ```rust
+    /// let bridge = hueclient::Bridge::for_ip([192u8, 168, 0, 4]);
+    /// ```
+    pub fn for_ip(ip: impl Into<std::net::IpAddr>) -> UnauthBridge {
```

Comment:
Should it be https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html ?

Reply:
I can use this one as well, but `ToSocketAddrs` (plural) returns an Iterator over zero or more addresses, so this would create some additional failure modes. I think that we can more easily get by using Into<SocketAddr>. I did write Into<IpAddr> which prohibits any ports from being specified, so I should probably still change this.

Reply:
Come to think of it, the ip address is returned by ssdp_probe, so it is probably correct.